tests/test_schema.py

```diff
@@ -567,3 +590,54 @@ def test_patchset_fail(datadir, patchset_file):
     patchset = json.load(open(datadir.joinpath(patchset_file)))
     with pytest.raises(pyhf.exceptions.InvalidSpecification):
         pyhf.schema.validate(patchset, 'patchset.json')
+
+
+@pytest.fixture
+def refresh_pyhf(monkeypatch):
+    global pyhf
+    modules_to_clear = [name for name in sys.modules if name.split('.')[0] == 'pyhf']
+    for module_name in modules_to_clear:
+        monkeypatch.delitem(sys.modules, module_name)
+    pyhf = importlib.import_module(pyhf.__name__)
+    return pyhf
```

Comment:
I like @kratsg's suggestion about not having to try to deal with `importlib` if it can be ignored, but could the equivalent effect be done with just `reload` like https://github.com/scikit-hep/pyhf/blob/97664bcfcef3b4a459a7f700b0703b40aec9b842/tests/test_scripts.py#L620-L625?

Reply:
Thanks for suggesting `mock.patch.dict`, it does look a little cleaner now.

However it seems leaving `pyhf` in `sys.modules` and `reload`ing it instead somehow causes the old cache to stick around after the import, so the test fails to catch the regression. This is a little mysterious to me. Maybe it is related to this line in the [docs](https://docs.python.org/3.7/library/importlib.html#importlib.reload):

> When a module is reloaded, its dictionary (containing the module’s global variables) is retained.

I would have expected the imports in `pyhf/__init__.py` to be re-interpreted during `reload`, since those are still missing from `sys.modules` at this point. But it seems not (perhaps bc. their local symbols are part of the retained dictionary for the  `pyhf` module?)

for future reference, this is the patch trying `reload`  that fails to catch the regression:
```diff
--- a/tests/test_schema.py
+++ b/tests/test_schema.py
@@ -611,10 +611,12 @@ def test_defs_always_cached(
     """

     global pyhf
-    modules_to_clear = [name for name in sys.modules if name.split('.')[0] == 'pyhf']
+    modules_to_clear = [
+        nom for nom in sys.modules if (nom.split('.')[0] == 'pyhf') and (nom != 'pyhf')
+    ]
     for module_name in modules_to_clear:
         del monkeypatched_sys_modules[module_name]
-    pyhf = importlib.import_module(pyhf.__name__)
+    pyhf = importlib.reload(pyhf)

     spec = {
         'channels': [
```